### PR TITLE
retworkx Paper V2

### PR DESCRIPTION
**Pull request by IvanIsCoding**
This is a rewrite of the retworkx paper. It has a completely rewritten Use Cases section, which is based on Qiskit's usage of retworkx. It also enchances the Statement of Need and Related Work sections and adds a Limitations section.

### CODE REVIEW

**Comment by IvanIsCoding** [comment]
Preview for easier reading:
[draftv2.pdf](https://github.com/Qiskit/retworkx/files/7833723/draftv2.pdf)

**Review by mtreinish — COMMENTED** [review]
I've done a quick skim over this to start. It looks good thanks for tackling this! I left a couple of high level comments/questions inline to start

File: paper.md
Comment (IvanIsCoding) [reply paper.md]: Thanks for catching this. I rephrased it to not imply all instructions are OpenQASM instructions, but I kept the cite to the OpenQASM paper as it is the best summary of what the instructions are.

File: paper.md
Comment (IvanIsCoding) [reply paper.md:67]: I took the number out of https://github.com/mtreinish/retworkx-comparison-benchmarks. Also, I took the freedom to create #550. We could use a summary of the results in the docs for questions like that

File: paper.md
Comment (IvanIsCoding) [reply paper.md:87]: Ah, I always thought we had VF2++ implemented. I will wait for the confirmation then

File: paper.md
Comment (IvanIsCoding) [reply paper.md]: I rephrased the section, I hope it is better. But after #248 and some of the review comments for the v1, I feel that "retworkx is not a drop-in replacement for NetworkX" should be on the paper

File: paper.md
Comment (georgios-ts) [reply paper.md:87]: Yeah, we only implement a simplified version of VF2++. In the matching order we do not take into consideration the node labels and also we do not implement a cutting rule, they proposed, based again in node labels.

Comment (IvanIsCoding) [reply paper.md]: Thanks for catching. The paper claims we implement some of the VF2++ heuristics now, which is accurate

**Comment by szhorvat** [comment]
Can you please avoid calling the data associated to nodes/edges "weights" in the paper? This is not standard usage. I expect readers will not understand it. I certainly found it very confusing when I encountered it in the documentation.

In standard usage, a "weight" is a number. Here you are referring to something different: an analog of what other graph libraries call "attributes" or "properties", which can in general be any kind of data. If I had to make a suggestion for naming, I'd say "associated data".

File: paper.md
Comment (szhorvat) [reply paper.md]: I still wonder if you made the networkx team aware of this library, and what they thought about the naming.

File: paper.md
Comment (mtreinish) [inline paper.md:89]: I would rename this section or integrate the contents here into an earlier section outlining what retworkx is. This seems out of place as a standalone section. I get what you're saying about it being important to call out in the paper to avoid potential confusion about it being a drop in replacement but having a standalone section titled limitations doesn't seem like the right fit for that intent. (especially because it really was a design decision not a limitation of the library).

**Review by mtreinish — COMMENTED** [review]
I think this LGTM, I'm happy with this draft after adjusting my one inline comment. But it'd be good to get a review from @georgios-ts and @nahumsa too before we merge this.

File: paper.md
Comment (nahumsa) [reply paper.md:67]: I agree that this should have a link or some reference.

File: paper.md
Comment (nahumsa) [inline paper.md:61]: Maybe could mention that it was generated using the `qiskit.generators` module as well.

File: paper.md
Comment (nahumsa) [inline paper.md:83]: I think that the reference here, should be the original NISQ paper written by Preskill, [reference](https://quantum-journal.org/papers/q-2018-08-06-79/).

File: paper.md
Comment (nahumsa) [inline paper.md:87]: I think that "SWAP instructions" are ok for people with knowledge in quantum computing, since this should be readable for a broader audience, I think that it might be better to say "swap qubits", or maybe "decrease the depth of the quantum circuit" (which is also more suited for the quantum computing audience).

**Review by nahumsa — COMMENTED** [review]
LGTM. I think that the paper now is much more concise and focused on a particular application, which I think that was what the journal wanted.
I have some questions/comments about some minor things.

File: paper.md
Comment (IvanIsCoding) [reply paper.md]: I changed it to "If they are, Qiskit uses VF2 mapping to map the qubits without increasing the depth of the circuit.".

File: paper.md
Comment (IvanIsCoding) [reply paper.md]: I cited both papers, the Bharti et al. paper has more details about qubit mapping so I kept it

**Review by nahumsa — APPROVED** [review]

**Review by georgios-ts — APPROVED** [review]
Nice!

File: paper.md
Comment (georgios-ts) [reply paper.md:89]: I would keep the section and call out here that there are some functionality gaps with NetworkX.

File: paper.md
Comment (georgios-ts) [inline paper.md:53]: we could add a comment that we also support also multigraphs.

File: paper.md
Comment (mtreinish) [reply paper.md:89]: Yeah, that works too, that would make this more about current limitations with the library (and one we'll hopefully address over time)

**Review by mtreinish — APPROVED** [review]
Thanks for updating, I think this LGTM now too. I still think there are little things, like @georgios-ts comment about mentioning multigraphs or having a link to the benchmarks when we mention the performance, but those are easy to tweak later

**Comment by IvanIsCoding** [comment]
I updated the pre-print on the arXiv, v2 should be announced on Monday

**Pull request merged**
